loopchain/blockchain/blocks/v1_0/block_verifier.py

```diff
@@ -1,25 +1,137 @@
+from loopchain.blockchain.votes import v0_1a
+from loopchain.crypto.signature import SignVerifier
 
 if TYPE_CHECKING:
-    from loopchain.blockchain.invoke_result import InvokeData, InvokePool
-    from loopchain.blockchain.blocks.v1_0.block import Block
+    from loopchain.blockchain.invoke_result import InvokePool
+    from loopchain.blockchain.blocks.v1_0.block import Block, BlockHeader
 
 
 class BlockVerifier(DataVerifier):
     version = "1.0"
 
-    def __init__(self, tx_versioner: TransactionVersioner, invoke_pool: 'InvokePool'):
+    def __init__(self, tx_versioner: TransactionVersioner, invoke_pool: 'InvokePool', reps_getter):
         self._invoke_pool: 'InvokePool' = invoke_pool
         self._tx_versioner: TransactionVersioner = tx_versioner
+        self._reps_getter = reps_getter
 
     async def verify(self, prev_data: 'Block', data: 'Block'):
-        self._do_invoke(data)
+        self._verify_version(data)
+        self._verify_common(prev_data, data)
+        self._verify_transactions(data)
+        self._verify_signature(data)
+        self._invoke_pool.invoke(data, self._tx_versioner)
 
-    def _do_invoke(self, block) -> 'InvokeData':
-        invoke_result = self._invoke_pool.invoke(block=block)
+    def _verify_version(self, block: 'Block'):
+        if block.header.version != self.version:
+            raise BlockVersionNotMatch(block.header.version, self.version,
+                                       f"The block version is incorrect. Block({block.header})")
 
-        return invoke_result
+    def _verify_common(self, prev_block: 'Block', block: 'Block'):
+        header: 'BlockHeader' = block.header
+
+        if header.timestamp is None:
+            raise RuntimeError(f"Block({header.height}, {header.hash.hex()} does not have timestamp.")
+        if header.height > 0 and header.prev_hash is None:
+            raise RuntimeError(f"Block({header.height}, {header.hash.hex()} does not have prev_hash.")
+
+        if header.height > 1:
+            prev_next_reps_hash = prev_block.header.revealed_next_reps_hash
+            if prev_next_reps_hash:
+                if prev_next_reps_hash != header.validators_hash:
+                    raise RuntimeError(
+                        f"Block({header.height}, {header.hash.hex()}, "
+                        f"RepsHash({header.validators_hash}), "
+                        f"Expected({prev_next_reps_hash}).")
+                prev_reps = self._reps_getter(prev_block.header.validators_hash)
+            else:
+                prev_reps = self._reps_getter(block.header.validators_hash)
+            self.verify_prev_votes(block, prev_reps)
+
+    def verify_prev_votes(self, block: 'Block', prev_reps: Sequence['ExternalAddress']):
+        prev_votes: 'BlockBody' = block.body.prev_votes
+
+        if prev_votes:
+            for vote in prev_votes:
+                if vote and vote.version <= '0.5':
```

Comment:
You intended: ?
- first `Block` of 1.0 contains Vote 0.5.

Reply:
Yes, for once I consider the case the first block of the `LFT` contains votes of a block by the `Siever`.